Auth and fb db

ShonOhana opened the pull request:
1. make login and register screen according to lectures. (No navigation implemented yet in this screens until we sit and decide)
2. create repositories and interactor to implement firebase services.
3. auth to firebase authetication functionallity and push to realtime database.

ShonOhana committed:
Update database operation

ShonOhana committed:
1. make login and register screen according to lectures. (No navigation implemented yet in this screens until we sit and decide)
2. create repositories and interactor to implement firebase services.
3. auth to firebase authetication functionallity and push to realtime database.

RazCoH commented on app/src/main/java/com/example/composetemplate/data/models/local_models/User.kt:
```diff
@@ -0,0 +1,15 @@
+@Serializable
+data class User(
+    var permission_type: String,
+    var full_name: String,
```
There is special reason to use snake case instead of camel case?

ShonOhana replied:
Note to my self: check if @Serialize... is good to push as body

RazCoH commented on app/src/main/java/com/example/composetemplate/presentation/activities/MainActivity.kt:
```diff
@@ -14,7 +17,28 @@ class MainActivity : ComponentActivity() {
         setContent {
             ComposeTemplateTheme {
-                Navigator().CreateAppNavigation()
+//                Navigator().CreateAppNavigation()
```
Now I understand what you meant. If we don't want the changes in this file, why push it? Alternatively, we can push just the changes that we need.

ShonOhana replied:
note for my self: it before pushing the master. We need to talk about architecture of screens and navigation before I develop it to the master

RazCoH commented on app/src/main/java/com/example/composetemplate/presentation/common/LoginScreenButton.kt:
```diff
@@ -0,0 +1,46 @@
+
+
+@Composable
+fun LoginScreenButton(
```
Just wondering, can't we create a general button component instead of a login specific one?

ShonOhana replied:
RAZ:::: you can check if you want to do base button, now its login componnent because compose button componnet already ask for color, shape, etc....

RazCoH commented on app/src/main/java/com/example/composetemplate/presentation/common/LoginTextField.kt:
```diff
@@ -0,0 +1,57 @@
+import com.example.composetemplate.presentation.screens.entry_screens.login.LoginScreenEnum
+
+@Composable
+fun LoginTextField(
```
It's looks like a complex component, but maybe here we can also make it more generic?

ShonOhana replied:
RAZ same concept as LoginScreenButton

RazCoH commented on app/src/main/java/com/example/composetemplate/presentation/screens/entry_screens/login/LoginScreen.kt:
```diff
@@ -0,0 +1,81 @@
+            modifier = modifier
+                .padding(top = 12.dp)
+                .clickable { onRegisterPagePageClick() },
+            text = "Don't have an account? Register! ",
```
We agreed to put texts in the constants file

ShonOhana replied:
I will do it, good notice

RazCoH commented on app/src/main/java/com/example/composetemplate/presentation/screens/entry_screens/login/LoginScreenEnum.kt:
```diff
@@ -0,0 +1,46 @@
+
+    fun getPlaceHolder(): String{
+        return when(this) {
+            EMAIL -> "Email"
```
Also here, put it in Constants file please.

ShonOhana replied:
I will do it, good notice

RazCoH commented on app/src/main/java/com/example/composetemplate/presentation/screens/entry_screens/AuthenticationScreen.kt:
```diff
@@ -7,7 +7,7 @@ import com.example.composetemplate.navigation.Navigator
 fun AuthenticationScreen(args: EntryScreens.Authentication, navigator: Navigator) {
     if (args.isAuthorized) {
-        LogInScreen(navigator, args.shouldShowAdd)
+        RegisterScreen(navigator, args.shouldShowAdd) //todo: move
```
Can you explain this? where this component should move?

ShonOhana replied:
It will change in the navigation mission before master

RazCoH commented on app/src/main/java/com/example/composetemplate/repositories/AuthInteractor.kt:
```diff
@@ -0,0 +1,56 @@
+ *  RECOMMENDATION: Add SHA1 key in firebase project settings, you need it for example, in google auth
+ */
+
+class AuthInteractor(
```
I would appreciate it if you could explain the role of this class to me in a Zoom meeting.Like what is the difference for us between an Interactor and a Repository? I see there is a LoginRepository, and this class is in the repositories package.

ShonOhana replied:
Shon todo: add explaind comment about the Intercator. Main Idea basically its to do more manipulate about logic that you cant do in the repo. like here, I needed to use network class that i Couldnt inject to repo. basically we need to do logic in the smallest layer we can. we can add another interactor package.

RazCoH commented on app/src/main/java/com/example/composetemplate/repositories/FirebaseDataSource.kt:
```diff
@@ -0,0 +1,69 @@
+import com.google.firebase.auth.FirebaseAuth
+import com.google.firebase.auth.auth
+
+class FirebaseDataSource() : AuthDBServiceable {
```
Just wondering if this shouldn't be a Manager class. Let's consider if DataSource is a component relevant in other cases as well. If it is, maybe we should create a new package for data sources.

ShonOhana replied:
same concept as AuthInteractor, make explained comment

RazCoH commented on app/src/main/java/com/example/composetemplate/repositories/LoginRepository.kt:
```diff
@@ -0,0 +1,57 @@
+/**
+ * This class contain the functionality of all login auth types
+ */
+class LoginRepository(
```
As I said let's talk about it and the itercator

ShonOhana replied:
same concept as AuthInteractor, make explained comment

RazCoH commented on app/src/main/java/com/example/composetemplate/utils/DateUtil.kt:
```diff
@@ -0,0 +1,21 @@
+class DateUtil {
+
+    companion object {
+        const val SERVER_DATE_FORMAT = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"  //"2021-04-26T10:00:00Z"
```
I think I asked that before in other PR, there is special reason for it to be companion object? I think it will be better to make it an extension function of Long

ShonOhana replied:
todo: Ask guy

RazCoH reviewed (changes requested):
Good Job, 
There are few comments about the code.
 But there are some things that I want to chat about (in the architecture level)

ShonOhana committed:
fixes the last pr comments.
all is left is to ask Guy about DateUtil and talk about navigation in the next branch (before master)

RazCoH commented on app/src/main/java/com/example/composetemplate/data/remote/base/BaseNetworking.kt:
```diff
@@ -18,6 +24,7 @@ Additionally, this class offers a preconfigured base client for immediate use. I
 @see Networking implementation */
 abstract class BaseNetworking : BaseServerClient {
 
+    @OptIn(ExperimentalSerializationApi::class)
```
is this annotation needed? If it is, for what reason?

ShonOhana committed:
add comment about @OptIn(ExperimentalSerializationApi::class)
